### PR TITLE
kubectl: Remove multiple Visit calls on the visitor source when computing resources to scale

### LINKED ISSUE #1183: kubectl scale doesn't work with piped file input
**What happened**:
While scaling deployments using piped input, it throws an error
`cat test_deployment.json | kubectl scale --replicas=0 -f -`
 _error: no objects passed to scale_

**What you expected to happen**:
kubectl scale should work from piped file (-f -) and it should scale those resources

**How to reproduce it (as minimally and precisely as possible)**:
1. create a simple nginx deployment and then Scale the nginx deployment using piped file input
```
kubectl create deployment nginx --image=nginx
kubectl get deploy nginx -o json  | kubectl scale --replicas=0 -f -
```

 ( it throws an error _error: no objects passed to scale_ )

**Anything else we need to know?**:
If we store the output of kubectl get deploy nginx -o json to a file and then use that filename, the scale command works fine
```
kubectl get deploy nginx -o json > test_deployment.json
kubectl scale --replicas=0 -f test_deployment.json
```

**Environment**:
- Kubernetes client and server versions (use `kubectl version`): v1.21.5(client)/v1.21.5-eks-bc4871b(server)
- Cloud provider or hardware configuration: EKS
- OS (e.g: `cat /etc/os-release`): MacOS ( 11.6 )
Screen Shot 2022-03-07 at 10 54 33 AM

### PR DESCRIPTION
**Pull request by harjas27**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
`RunScale()` contains two `Visit(fn VisitorFunc)` calls. When the input for the `scale` command is from pipe, the second call does not result in any `info` which results in error
Adding the fix to store the output from the first call into a `infos` variable and reusing it the second time
#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/1183

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/scale/scale.go [hunk 1]
```diff
@@ -264,7 +258,8 @@ func (o *ScaleOptions) RunScale() error {
 		}
 
 		return o.PrintObj(info.Object, o.Out)
-	})
+
```
Comment (knight42) [inline staging/src/k8s.io/kubectl/pkg/cmd/scale/scale.go:261]: nit: we could drop this empty line.

**Review by knight42 — APPROVED** [review]
/lgtm
/release-note none

@harjas27 Hi! Thanks for your contribution! Could you please add an e2e test to https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubectl/kubectl.go to prevent possible regression in the future?

File: staging/src/k8s.io/kubectl/pkg/cmd/scale/scale.go [hunk 2]
```diff
@@ -264,7 +258,8 @@ func (o *ScaleOptions) RunScale() error {
 		}
 
 		return o.PrintObj(info.Object, o.Out)
```
Comment (knight42) [inline staging/src/k8s.io/kubectl/pkg/cmd/scale/scale.go:260]: Ooops.. Now that we are iterating `infos`, we could not simply return inside the for loop. 

As an alternative, we could wrap the original visit logic inside a new funcion, for instance:
```
visit := func(info *resource.Info) error {
  ...
}
for _, info := range infos {
  err := visit(info)
  if err != nil {
    return err
  }
}
```

File: staging/src/k8s.io/kubectl/pkg/cmd/scale/scale.go [hunk 3]
```diff
@@ -235,10 +232,7 @@ func (o *ScaleOptions) RunScale() error {
 	}
 
 	counter := 0
```
Comment (knight42) [inline staging/src/k8s.io/kubectl/pkg/cmd/scale/scale.go:234]: We could drop this counter and check the length of `infos` directly.

**Comment by pacoxu** [comment]
```

test-cmd: run_recursive_resources_tests expand_less	6s
{Script Error ScriptError error: error validating "hack/testdata/recursive/rc/rc/busybox-broken.yaml": error validating data: kind not set; if you choose to ignore these errors, turn validation off with --validate=false
!!! [0621 07:42:16] Call tree:
!!! [0621 07:42:16]  1: /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../test/cmd/generic-resources.sh:346 kube::test::get_object_assert(...)
!!! [0621 07:42:16]  2: /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../third_party/forked/shell2junit/sh2ju.sh:47 run_recursive_resources_tests(...)
!!! [0621 07:42:16]  3: /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../third_party/forked/shell2junit/sh2ju.sh:112 eVal(...)
!!! [0621 07:42:16]  4: /home/prow/go/src/k8s.io/kubernetes/test/cmd/legacy-script.sh:138 juLog(...)
!!! [0621 07:42:16]  5: /home/prow/go/src/k8s.io/kubernetes/test/cmd/legacy-script.sh:614 record_command(...)
!!! [0621 07:42:16]  6: hack/make-rules/test-cmd.sh:194 runTests(...)}
```
Integration test is broken. Would you fix it?

**Pull request closed without merging**